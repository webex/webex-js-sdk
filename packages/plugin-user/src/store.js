/**!
 *
 * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
 */

import {patterns} from '@ciscospark/common';

const usersByEmail = new WeakMap();
const usersById = new WeakMap();

export default class UserStore {
  constructor() {
    usersByEmail.set(this, new Map());
    usersById.set(this, new Map());
  }

  add(user) {
    if (!user.id) {
      return Promise.reject(`\`user.id\` is required`);
    }

    if (!user.emailAddress) {
      return Promise.reject(`\`user.emailAddress\` is required`);
    }

    const p1 = this.getById(user.id)
      .then((u) => usersById.get(this).set(user.id, Object.assign({}, u, user)))
      .catch(() => usersById.get(this).set(user.id, Object.assign({}, user)));

    const p2 = this.getByEmail(user.emailAddress)
      .then((u) => usersByEmail.get(this).set(user.emailAddress, Object.assign({}, u, user)))
      .catch(() => usersByEmail.get(this).set(user.emailAddress, Object.assign({}, user)));

    return Promise.all([p1, p2]);
  }

  get(id) {
    if (patterns.uuid.test(id)) {
      return this.getById(id);
    }

    if (patterns.email.test(id)) {
      return this.getByEmail(id);
    }

    return Promise.reject(`\`id\` does not appear to be a valid user identifier`);
  }

  getById(id) {
    const ret = usersById.get(this).get(id);
    if (ret) {
      return Promise.resolve(ret);
    }
    return Promise.reject(new Error(`No user found identified by specified identifier`));
  }

  getByEmail(email) {
    const ret = usersByEmail.get(this).get(email);
    if (ret) {
      return Promise.resolve(ret);
    }
    return Promise.reject(new Error(`No user found identified by specified identifier`));
  }
}
