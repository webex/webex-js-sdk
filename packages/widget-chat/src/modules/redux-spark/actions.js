export const UPDATE_SPARK_STATE = `UPDATE_SPARK_STATE`;
export function updateSparkState(state) {
  return {
    type: UPDATE_SPARK_STATE,
    state
  };
}

export function connectToMercury(spark) {
  return (dispatch) => {
    spark.mercury.connect()
      .then(dispatch(updateSparkState({connected: true, connecting: false})));
    dispatch(updateSparkState({connecting: true}));
  };
}
