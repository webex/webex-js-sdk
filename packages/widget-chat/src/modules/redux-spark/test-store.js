// Store for testing
import {
  applyMiddleware,
  combineReducers,
  compose,
  createStore
} from 'redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Map} from 'immutable';
import reducers from './reducers';

jest.mock(`./spark`);
const {createSpark} = require(`./spark`);

export function createMockStore() {
  const mockStore = configureMockStore([thunk]);
  return mockStore(new Map({
    status: new Map({
      authenticated: false,
      authenticating: false,
      registered: false,
      registering: false,
      connected: false,
      connecting: false
    }),
    spark: createSpark()
  }));
}

const enhancers = [
  applyMiddleware(
    thunk)
];

export const store = createStore(
  combineReducers({
    spark: reducers
  }),
  compose(...enhancers)
);
