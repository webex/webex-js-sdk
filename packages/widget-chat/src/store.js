import {applyMiddleware, combineReducers, compose, createStore} from 'redux';

import thunk from 'redux-thunk';
import createLogger from 'redux-logger';
import {user, conversation} from './reducers';
import sparkReducer from './modules/redux-spark/reducers';

const devtools = window.devToolsExtension || (() => (noop) => noop);

const enhancers = [
  applyMiddleware(
    thunk,
    createLogger({
      level: `info`,
      duration: true,
      collapsed: false,
      logErrors: process.env.NODE_ENV === `production`
    })
  ),
  devtools()
];

export default createStore(
  combineReducers({
    user,
    conversation,
    spark: sparkReducer
  }),
  compose(...enhancers)
);
