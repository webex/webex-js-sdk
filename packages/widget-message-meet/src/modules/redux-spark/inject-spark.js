import React, {Component, PropTypes} from 'react';

import SparkComponent from './component';

function getDisplayName(C) {
  return C.displayName || C.name || `C`;
}

export default function injectSpark(WrappedComponent, options = {}) {
  const {
    withRef = false
  } = options;

  class InjectSpark extends Component {

    shouldComponentUpdate(nextProps) {
      return nextProps !== this.props;
    }

    getWrappedInstance() {
      // eslint-disable-next-line react/no-string-refs
      return this.refs.wrappedInstance;
    }

    render() {
      return (
        <div>
          <SparkComponent accessToken={this.props.accessToken} />
          <WrappedComponent
            {...this.props}
            ref={withRef ? `wrappedInstance` : null}
          />
        </div>
      );
    }
  }

  InjectSpark.propTypes = {
    accessToken: PropTypes.string
  };

  InjectSpark.displayName = `InjectSpark(${getDisplayName(WrappedComponent)})`;
  InjectSpark.WrappedComponent = WrappedComponent;

  return InjectSpark;
}
