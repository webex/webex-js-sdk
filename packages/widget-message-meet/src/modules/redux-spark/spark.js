/* eslint camelcase: [0] */

import '@ciscospark/plugin-avatar';
import '@ciscospark/plugin-conversation';
import '@ciscospark/plugin-flag';
import Spark from '@ciscospark/spark-core';
import LocalStorageStoreAdapter from '@ciscospark/storage-adapter-local-storage';

export function createSpark(accessToken) {
  const credentials = JSON.parse(localStorage.getItem(`credentials`) || false);
  return new Spark({
    credentials: {
      authorization: {
        access_token: accessToken || credentials.access_token
      }
    },
    config: {
      conversation: {
        allowedInboundTags: {
          'spark-mention': [`data-object-type`, `data-object-id`, `data-object-url`],
          a: [`href`],
          b: [],
          blockquote: [`class`],
          strong: [],
          i: [],
          em: [],
          pre: [],
          code: [],
          br: [],
          hr: [],
          p: [],
          ul: [],
          ol: [],
          li: [],
          h1: [],
          h2: [],
          h3: []
        },
        allowedOutboundTags: {
          'spark-mention': [`data-object-type`, `data-object-id`, `data-object-url`],
          a: [`href`],
          b: [],
          blockquote: [`class`],
          strong: [],
          i: [],
          em: [],
          pre: [],
          code: [],
          br: [],
          hr: [],
          p: [],
          ul: [],
          ol: [],
          li: [],
          h1: [],
          h2: [],
          h3: []
        }
      },
      storage: {
        boundedAdapter: new LocalStorageStoreAdapter(`ciscospark-embedded`)
      }
    }
  });
}
