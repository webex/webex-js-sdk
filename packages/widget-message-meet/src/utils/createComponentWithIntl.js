import React from 'react';
import renderer from 'react-test-renderer';
import {IntlProvider} from 'react-intl';

function createComponentWithIntl(children, props = {locale: `en`}) {
  return renderer.create(
    <IntlProvider {...props}>
      {children}
    </IntlProvider>
  );
}

export default createComponentWithIntl;
