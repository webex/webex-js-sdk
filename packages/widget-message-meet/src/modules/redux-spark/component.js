import {Component, PropTypes} from 'react';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';

import {
  connectToMercury,
  updateSparkState,
  registerDevice,
  storeSparkInstance
} from './actions';

import {createSpark} from './spark';

class SparkComponent extends Component {

  componentDidMount() {
    const {
      accessToken
    } = this.props;

    const props = this.props;
    let spark = props.spark.get(`spark`);

    if (!spark) {
      spark = createSpark(accessToken);
      props.storeSparkInstance(spark);
    }

    spark.listenToAndRun(spark, `change:isAuthenticated`, () => {
      props.updateSparkState({authenticated: spark.isAuthenticated});
    });

    spark.listenToAndRun(spark, `change:isAuthenticating`, () => {
      props.updateSparkState({authenticating: spark.isAuthenticating});
    });

    spark.mercury.listenToAndRun(spark.mercury, `change:connected`, () => {
      props.updateSparkState({connected: spark.mercury.connected});
    });

    if (props.authenticated && !props.registered && !props.registering) {
      props.registerDevice(spark);
    }
  }

  componentWillReceiveProps(nextProps) {
    const {
      connected,
      connecting,
      authenticated,
      registered,
      registering
    } = nextProps.spark.get(`status`).toJS();
    const spark = nextProps.spark.get(`spark`);

    if (authenticated && !registered && !registering) {
      nextProps.registerDevice(spark);
    }
    else if (registered && authenticated && !connected && !connecting) {
      nextProps.connectToMercury(spark);
    }
  }

  shouldComponentUpdate(nextProps) {
    const props = this.props;
    return nextProps !== props;
  }

  render() {
    return null;
  }
}

SparkComponent.propTypes = {
  accessToken: PropTypes.string
};

export default connect(
  (state) => ({
    spark: state.spark
  }),
  (dispatch) => bindActionCreators({
    connectToMercury,
    updateSparkState,
    registerDevice,
    storeSparkInstance
  }, dispatch)
)(SparkComponent);
