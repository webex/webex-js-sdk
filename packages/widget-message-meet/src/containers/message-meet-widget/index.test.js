import React from 'react';
import {Provider} from 'react-redux';

import createComponentWithIntl from '../../utils/createComponentWithIntl';

import {store as sparkStore} from '../../modules/redux-spark/test-store';
import store from '../../store';
import MessageMeetWidget from '.';

store.spark = sparkStore;

describe(`MessageMeetWidget`, () => {
  it(`renders properly`, () => {
    const accessToken = process.env.CISCOSPARK_ACCESS_TOKEN;
    const component = createComponentWithIntl(
      <Provider store={store}>
        <MessageMeetWidget accessToken={accessToken} userId="sparky@ciscospark.com" />
      </Provider>
    );

    const widget = component.toJSON();

    expect(widget).toMatchSnapshot();
  });
});
