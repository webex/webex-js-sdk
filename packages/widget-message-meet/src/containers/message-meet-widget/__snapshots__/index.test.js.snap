exports[`MessageMeetWidget renders properly 1`] = `
<div>
  <div
    className="widget-chat">
    <div
      className="banner" />
    <div
      className="loading">
      <span>
        Connecting
      </span>
      ...
      <div
        className="spinner-container">
        <div
          className="spinner" />
      </div>
    </div>
  </div>
</div>
`;
